File: multi-lang/netcore/ChainedReader.cs
namespace QQZeng.Qzdb;

using System.Collections.Generic;
using System.Collections.ObjectModel;

/// <summary>Combines multiple <see cref="QzdbReader"/> instances into one logical query surface, with Fallback / Merge / MergeOverride strategies.</summary>
public sealed class ChainedReader : IDisposable
{
    /// <summary>Strategy used to combine results from the chained readers.</summary>
    public enum Mode
    {
        /// <summary>Returns the first reader's hit, in order; later readers are consulted only on a miss.</summary>
        Fallback,
        /// <summary>Merges all hits; a later reader fills only fields left empty by earlier readers (earlier wins).</summary>
        Merge,
        /// <summary>Merges all hits; a later reader's value overrides an earlier reader's value (later wins).</summary>
        MergeOverride
    }

    private readonly ReadOnlyCollection<QzdbReader> _readers;
    private readonly Mode _mode;

    private ChainedReader(QzdbReader[] readers, Mode mode)
    {
        ArgumentNullException.ThrowIfNull(readers);
        if (readers.Any(r => r == null)) throw new ArgumentException("Readers cannot contain null", nameof(readers));
        _readers = Array.AsReadOnly((QzdbReader[])readers.Clone());
        _mode = mode;
    }

    /// <summary>Creates a Fallback-mode chained reader over the given readers.</summary>
    public static ChainedReader Chain(params QzdbReader[] readers) => new(readers, Mode.Fallback);
    /// <summary>Creates a Merge-mode chained reader over the given readers.</summary>
    public static ChainedReader ChainMerge(params QzdbReader[] readers) => new(readers, Mode.Merge);
    /// <summary>Creates a MergeOverride-mode chained reader over the given readers.</summary>
    public static ChainedReader ChainMergeOverride(params QzdbReader[] readers) => new(readers, Mode.MergeOverride);

    /// <summary>Queries the chain for an IP string (per-mode semantics; see <see cref="QzdbReader.Find(string)"/>).</summary>
    public GeoInfo? Find(string ipStr) => Find(ipStr.AsSpan());

    /// <summary>Queries the chain for an IP character span (per-mode semantics).</summary>
    public GeoInfo? Find(ReadOnlySpan<char> ipSpan)
    {
        if (_mode == Mode.Fallback)
        {
            for (int i = 0; i < _readers.Count; i++)
            {
                var res = _readers[i].Find(ipSpan);
                if (res != null) return res;
            }
            return null;
        }

        var indexes = new Dictionary<string, int>(StringComparer.Ordinal);
        var names = new List<string>();
        var values = new List<string>();
        for (int i = 0; i < _readers.Count; i++)
        {
            var info = _readers[i].Find(ipSpan);
            if (info != null) MergeInfo(indexes, names, values, info);
        }
        if (names.Count == 0) return null;
        var nameArray = names.ToArray();
        return new GeoInfo(nameArray, values.ToArray(), GeoInfo.BuildNormalizedMap(nameArray), null, takeOwnership: true);
    }

    /// <summary>Queries the chain for a System.Net.IPAddress (per-mode semantics).</summary>
    public GeoInfo? Find(System.Net.IPAddress address)
    {
        ArgumentNullException.ThrowIfNull(address);
        Span<byte> bytes = stackalloc byte[16];
        if (address.TryWriteBytes(bytes, out int written))
        {
            return Find(bytes[..written]);
        }
        return null;
    }

    /// <summary>Queries the chain for raw IP bytes (4 or 16); per-mode semantics.</summary>
    public GeoInfo? Find(ReadOnlySpan<byte> ipBytes)
    {
        if (_mode == Mode.Fallback)
        {
            for (int i = 0; i < _readers.Count; i++)
            {
                var res = _readers[i].Find(ipBytes);
                if (res != null) return res;
            }
            return null;
        }

        var indexes = new Dictionary<string, int>(StringComparer.Ordinal);
        var names = new List<string>();
        var values = new List<string>();
        for (int i = 0; i < _readers.Count; i++)
        {
            var info = _readers[i].Find(ipBytes);
            if (info != null) MergeInfo(indexes, names, values, info);
        }
        if (names.Count == 0) return null;
        var nameArray = names.ToArray();
        return new GeoInfo(nameArray, values.ToArray(), GeoInfo.BuildNormalizedMap(nameArray), null, takeOwnership: true);
    }

    /// <summary>Queries the chain for an IPv4 uint (per-mode semantics).</summary>
    public GeoInfo? FindUint(uint ipInt)
    {
        if (_mode == Mode.Fallback)
        {
            for (int i = 0; i < _readers.Count; i++)
            {
                var res = _readers[i].FindUint(ipInt);
                if (res != null) return res;
            }
            return null;
        }

        var indexes = new Dictionary<string, int>(StringComparer.Ordinal);
        var names = new List<string>();
        var values = new List<string>();
        for (int i = 0; i < _readers.Count; i++)
        {
            var info = _readers[i].FindUint(ipInt);
            if (info != null) MergeInfo(indexes, names, values, info);
        }
        if (names.Count == 0) return null;
        var nameArray = names.ToArray();
        return new GeoInfo(nameArray, values.ToArray(), GeoInfo.BuildNormalizedMap(nameArray), null, takeOwnership: true);
    }

    /// <summary>Queries the chain for IP bytes; a null argument throws QzdbException(InvalidIp).</summary>
    public GeoInfo? FindBytes(byte[]? ipBytes)
    {
        if (ipBytes == null) throw new QzdbException(ErrorCode.InvalidIp, "IP bytes cannot be null");
        return Find(ipBytes.AsSpan());
    }

    /// <summary>Queries the chain for selected fields of an IP string (per-mode semantics).</summary>
    public GeoInfo? FindFields(string ipStr, string[]? fields) => FindFields(ipStr.AsSpan(), fields);

    /// <summary>Queries the chain for selected fields of an IP span (per-mode semantics).</summary>
    public GeoInfo? FindFields(ReadOnlySpan<char> ipSpan, string[]? fields)
    {
        if (_mode == Mode.Fallback)
        {
            for (int i = 0; i < _readers.Count; i++)
            {
                var res = _readers[i].FindFields(ipSpan, fields);
                if (res != null) return res;
            }
            return null;
        }

        var indexes = new Dictionary<string, int>(StringComparer.Ordinal);
        var names = new List<string>();
        var values = new List<string>();
        for (int i = 0; i < _readers.Count; i++)
        {
            var info = _readers[i].FindFields(ipSpan, fields);
            if (info != null) MergeInfo(indexes, names, values, info);
        }
        if (names.Count == 0) return null;
        var nameArray = names.ToArray();
        return new GeoInfo(nameArray, values.ToArray(), GeoInfo.BuildNormalizedMap(nameArray), null, takeOwnership: true);
    }

    /// <summary>Batch query over IP strings; each result carries its three-state semantics.</summary>
    public BatchResult[] FindBatch(string[] ipStrs)
    {
        ArgumentNullException.ThrowIfNull(ipStrs);
        var result = new BatchResult[ipStrs.Length];
        for (int i = 0; i < ipStrs.Length; i++) result[i] = FindResult(ipStrs[i]);
        return result;
    }

    /// <summary>Batch query over an enumerable of IP strings.</summary>
    public BatchResult[] FindBatch(IEnumerable<string> ipStrs) => FindBatch(ipStrs?.ToArray() ?? throw new ArgumentNullException(nameof(ipStrs)));

    /// <summary>Batch query over IP strings, resolving only the given fields.</summary>
    public BatchResult[] FindBatchFields(string[] ipStrs, string[]? fields)
    {
        ArgumentNullException.ThrowIfNull(ipStrs);
        var result = new BatchResult[ipStrs.Length];
        for (int i = 0; i < ipStrs.Length; i++) result[i] = LookupResult(ipStrs[i], fields);
        return result;
    }

    /// <summary>Batch query over enumerables of IP strings and fields.</summary>
    public BatchResult[] FindBatchFields(IEnumerable<string> ipStrs, IEnumerable<string>? fields) =>
        FindBatchFields(ipStrs?.ToArray() ?? throw new ArgumentNullException(nameof(ipStrs)), fields?.ToArray());

    /// <summary>Lazily streams batch results for an enumerable of IP strings.</summary>
    public IEnumerable<BatchResult> FindStream(IEnumerable<string> ipStrs)
    {
        if (ipStrs == null) yield break;
        foreach (var ip in ipStrs) yield return FindResult(ip);
    }

    private BatchResult FindResult(string ip)
    {
        try { return new BatchResult(Find(ip), null, ip); }
        catch (QzdbException e) { return new BatchResult(null, e, ip); }
    }

    private BatchResult LookupResult(string ip, string[]? fields)
    {
        try { return new BatchResult(FindFields(ip, fields), null, ip); }
        catch (QzdbException e) { return new BatchResult(null, e, ip); }
    }



    private void MergeInfo(Dictionary<string, int> indexes, List<string> names, List<string> values, GeoInfo info)
    {
        var fields = info.FieldNames;
        var vals = info.Values;
        for (int i = 0; i < fields.Length; i++)
        {
            var f = fields[i];
            var v = i < vals.Length && vals[i] != null ? vals[i] : "";
            var key = GeoInfo.NormalizeKey(f);
            if (!indexes.TryGetValue(key, out var index))
            {
                indexes[key] = names.Count;
                names.Add(f);
                values.Add(v);
            }
            else if (_mode == Mode.Merge && string.IsNullOrEmpty(values[index]) && !string.IsNullOrEmpty(v))
            {
                values[index] = v;
            }
            else if (_mode == Mode.MergeOverride)
                values[index] = v;
        }
    }

    /// <summary>Editions of the underlying readers, in chain order.</summary>
    public string[] Editions => _readers.Select(r => r.Edition).ToArray();
    /// <summary>Scopes of the underlying readers, in chain order.</summary>
    public string[] Scopes => _readers.Select(r => r.Scope).ToArray();
    /// <summary>Data months of the underlying readers, in chain order.</summary>
    public string[] DataMonths => _readers.Select(r => r.DataMonth).ToArray();
    /// <summary>Defensive copy of the underlying readers, in chain order.</summary>
    public QzdbReader[] Readers => _readers.ToArray();

    /// <summary>Returns <see cref="Editions"/>.</summary>
    public string[] GetEditions() => Editions;
    /// <summary>Returns <see cref="Scopes"/>.</summary>
    public string[] GetScopes() => Scopes;
    /// <summary>Returns <see cref="DataMonths"/>.</summary>
    public string[] GetDataMonths() => DataMonths;

    /// <summary>Releases only the aggregation state; does NOT close the underlying readers (per API spec §9.4).</summary>
    public void Dispose()
    {
        foreach (var r in _readers) GC.KeepAlive(r);
    }
}

File: ip-qzdb-sdk/java/src/main/java/com/qqzeng/qzdb/ChainedReader.java
package com.qqzeng.qzdb;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * 复合联合查询类 (ChainedReader)
 * <p>
 * 支持将多个 QzdbReader 组合（例如“国内精华版 + 全球旗舰版”），提供 Fallback 备性退避与 Merge 字段自动拼接。
 */
public final class ChainedReader {

    public enum Mode {
        /**
         * Fallback 备性模式：按链顺序依次查找，首个命中结果即返回
         */
        FALLBACK,

        /**
         * Merge 拼接模式 (默认先注册者优先)：依次查询所有 Reader，将字段自动合并拼接为单个 GeoInfo
         */
        MERGE,

        /**
         * Merge 拼接模式 (后注册者覆盖)：后注册库的非空字段值覆盖先注册库
         */
        MERGE_OVERRIDE
    }

    private final List<QzdbReader> readers;
    private final Mode mode;

    private ChainedReader(List<QzdbReader> readers, Mode mode) {
        if (readers == null || readers.isEmpty()) {
            throw new QzdbException(ErrorCode.INVALID_PARAM, "ChainedReader requires at least one QzdbReader");
        }
        this.readers = List.copyOf(readers);
        this.mode = mode;
    }

    public static ChainedReader chain(QzdbReader... readers) {
        return new ChainedReader(Arrays.asList(readers), Mode.FALLBACK);
    }

    public static ChainedReader chainMerge(QzdbReader... readers) {
        return new ChainedReader(Arrays.asList(readers), Mode.MERGE);
    }

    public static ChainedReader chainMergeOverride(QzdbReader... readers) {
        return new ChainedReader(Arrays.asList(readers), Mode.MERGE_OVERRIDE);
    }

    // =========================================================================
    // 查询 API 矩阵
    //
    // 三个单条查询入口（find / findUint / findBytes）共用同一套「按模式遍历 reader」的
    // 内核 resolve()，差别只在传入的单库查询函数。此前 MERGE / MERGE_OVERRIDE 模式下，
    // findUint 会先 String.format 成点分文本、findBytes 会先经 InetAddress 转成文本，
    // 再交给每个库重新解析一遍；现在直接调用各库的原生入口，省掉格式化与二次解析。
    // =========================================================================

    /** 单个 reader 上的一次查询。 */
    @FunctionalInterface
    private interface Lookup {
        Optional<GeoInfo> apply(QzdbReader reader);
    }

    private Optional<GeoInfo> resolve(Lookup lookup) {
        return switch (mode) {
            case FALLBACK -> fallback(lookup);
            case MERGE, MERGE_OVERRIDE -> merge(lookup);
        };
    }

    /** FALLBACK：首个命中即返回。输入格式错误立即终止，其余加载/内部错误跳过该库。 */
    private Optional<GeoInfo> fallback(Lookup lookup) {
        for (QzdbReader reader : readers) {
            try {
                Optional<GeoInfo> res = lookup.apply(reader);
                if (res.isPresent()) {
                    return res;
                }
            } catch (QzdbException e) {
                if (e.getErrorCode() == ErrorCode.INVALID_IP) {
                    throw e; // 输入格式错误立终止
                }
            }
        }
        return Optional.empty();
    }

    /** MERGE / MERGE_OVERRIDE：逐库字段级合并（字段序为首次出现序）。 */
    private Optional<GeoInfo> merge(Lookup lookup) {
        Map<String, String> mergedMap = new LinkedHashMap<>();

        for (QzdbReader reader : readers) {
            try {
                Optional<GeoInfo> res = lookup.apply(reader);
                if (res.isPresent()) {
                    GeoInfo info = res.get();
                    // 只读遍历，无需拷贝（公共 API 的 clone 语义仍由 fieldNames()/values() 保持）
                    String[] fields = info.fieldNamesRaw();
                    String[] vals = info.valuesRaw();

                    for (int i = 0; i < fields.length; i++) {
                        String f = fields[i];
                        String v = (i < vals.length && vals[i] != null) ? vals[i] : "";

                        if (mode == Mode.MERGE) {
                            // 先注册者优先：先注册库的非空值不被覆盖；
                            // 先注册库该字段缺失/为空时，才用后面库的值补上（规范 §9.1）
                            mergedMap.merge(f, v, (old, cur) -> old.isEmpty() ? cur : old);
                        } else if (!v.isEmpty() || !mergedMap.containsKey(f)) {
                            // 后注册者覆盖：后注册库的非空值覆盖先注册库
                            mergedMap.put(f, v);
                        }
                    }
                }
            } catch (QzdbException e) {
                if (e.getErrorCode() == ErrorCode.INVALID_IP) {
                    throw e;
                }
            }
        }

        if (mergedMap.isEmpty()) {
            return Optional.empty();
        }

        String[] fieldNames = mergedMap.keySet().toArray(new String[0]);
        String[] values = mergedMap.values().toArray(new String[0]);
        return Optional.of(new GeoInfo(fieldNames, values));
    }

    public Optional<GeoInfo> find(String ipStr) {
        return resolve(reader -> reader.find(ipStr));
    }

    public Optional<GeoInfo> findUint(int ipInt) {
        return resolve(reader -> reader.findUint(ipInt));
    }

    public Optional<GeoInfo> findBytes(byte[] ip16) {
        return resolve(reader -> reader.findBytes(ip16));
    }

    public Optional<GeoInfo> findFields(String ipStr, String[] fields) {
        Optional<GeoInfo> full = find(ipStr);
        if (full.isEmpty() || fields == null || fields.length == 0) {
            return full;
        }

        GeoInfo fullInfo = full.get();
        String[] values = new String[fields.length];
        for (int i = 0; i < fields.length; i++) {
            values[i] = fullInfo.get(fields[i]);
        }
        return Optional.of(new GeoInfo(fields, values));
    }

    public List<BatchResult> findBatch(List<String> ips) {
        if (ips == null) return Collections.emptyList();
        List<BatchResult> results = new ArrayList<>(ips.size());
        for (String ip : ips) {
            try {
                Optional<GeoInfo> info = find(ip);
                results.add(new BatchResult(ip, info, null));
            } catch (QzdbException e) {
                results.add(new BatchResult(ip, Optional.empty(), e));
            }
        }
        return results;
    }

    public List<BatchResult> findBatchFields(List<String> ips, String[] fields) {
        if (ips == null) return Collections.emptyList();
        List<BatchResult> results = new ArrayList<>(ips.size());
        for (String ip : ips) {
            try {
                Optional<GeoInfo> info = findFields(ip, fields);
                results.add(new BatchResult(ip, info, null));
            } catch (QzdbException e) {
                results.add(new BatchResult(ip, Optional.empty(), e));
            }
        }
        return results;
    }

    public java.util.stream.Stream<BatchResult> findStream(java.util.stream.Stream<String> ips) {
        if (ips == null) return java.util.stream.Stream.empty();
        return ips.map(ip -> {
            try {
                Optional<GeoInfo> info = find(ip);
                return new BatchResult(ip, info, null);
            } catch (QzdbException e) {
                return new BatchResult(ip, Optional.empty(), e);
            }
        });
    }

    // =========================================================================
    // 元信息聚合 API
    // =========================================================================

    public String[] editions() {
        return readers.stream().map(QzdbReader::getEdition).toArray(String[]::new);
    }

    public String[] scopes() {
        return readers.stream().map(QzdbReader::getScope).toArray(String[]::new);
    }

    public String[] dataMonths() {
        return readers.stream().map(QzdbReader::getDataMonth).toArray(String[]::new);
    }

    public List<QzdbReader> readers() {
        return readers;
    }
}
